docs/decisions/0024-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,366 @@
+
+## Out of scope
+
+1. We provide an example on how to send telemetry to Application Insights. Although other telemetry service options are supported technically, we will not cover possible ways of setting them up in this ADR.
+2. This ADR does not seek to modify the current instrumentation design in SK.
+3. We do not consider services that do not return token usage.
+
+## Decision Drivers
+
+- The framework should be telemetry service agnostic.
+- Enabling and disabling specific telemetry items should require minimum effort.
+
+- The following metrics should be emitted by SK:
+  - Input token usage for prompt
+    - Description: A prompt is the smallest unit that consumes tokens. A function can consist of multiple prompts.
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Service ID | Model ID | Value |
+      |---|---|---|---|---|
+      | Function | WritePoem | | GPT-3.5-Turbo | 40
+      | Function | TellJoke | | GPT-4 | 50
+      | Function | WriteAndTellJoke | | GPT-3.5-Turbo | 30
+      | Planner | CreateHandlebarsPlan | | GPT-3.5-Turbo | 100
+  - Output token usage for prompt
+    - Description: A prompt is the smallest unit that consumes tokens. A function can consist of multiple prompts.
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Service ID | Model ID | Value |
+      |---|---|---|---|---|
+      | Function | WritePoem | | GPT-3.5-Turbo | 40
+      | Function | TellJoke | | GPT-4 | 50
+      | Function | WriteAndTellJoke | | GPT-3.5-Turbo | 30
+      | Planner | CreateHandlebarsPlan | | GPT-3.5-Turbo | 100
+  - Model request time for prompt
+    - Description: A prompt is the smallest unit that consumes tokens. A function can consist of multiple prompts.
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Service ID | Model ID | Value |
+      |---|---|---|---|---|
+      | Function | WritePoem | | GPT-3.5-Turbo | 0.5
+      | Function | TellJoke | | GPT-4 | 0.5m
+      | Function | WriteAndTellJoke | | GPT-3.5-Turbo | 0.3m
+      | Planner | CreateHandlebarsPlan | | GPT-3.5-Turbo | 1m
+  - Aggregated execution time for functions
+    - Description: A function can consist of multiple prompts. The execution time of a function is the duration from start to end of a function's `invoke` call.
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Value |
+      |---|---|---|
+      | Function | WritePoem | 1m
+      | Function | TellJoke | 1m
+      | Function | WriteAndTellJoke | 1.5m
+      | Planner | CreateHandlebarsPlan | 2m
+  - Success count for planners
+    - Description: A planner run is considered successful when it generates a valid plan.
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
```

Comment:
Is there a Plan name we can include also so we know if it's just specific plans or all plans that are failing

Reply:
By plan name, do you mean an identifier of a specific plan instance of say the Handlebars Plan?